FatPkg/EnhancedFatDxe/Fat.h

```diff
@@ -84,6 +84,19 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #define FAT_FATCACHE_GROUP_MIN_COUNT      1
 #define FAT_FATCACHE_GROUP_MAX_COUNT      16
 
+// For cache block bits, use CPU native size
+#define DIRTY_BLOCKS          UINTN
+#define BITS_PER_BYTE         8
+#define DIRTY_BITS_PER_BLOCK  (sizeof (DIRTY_BLOCKS) * BITS_PER_BYTE)
+
+// largest cache line (64KB) / MinLbaSize (512) = 128 bits

```

Comment:
Where does 64KB value come from?  Also the LBA Size depends on the attributes of the underlying Block I/O Protocol that can be read from Block I/O.  Why use assumption of 512 here?  I could be possible for a Block I/O Protocol to have a block size smaller than 512.

Reply:
These are definitions that already exist in Fat.h: https://github.com/tianocore/edk2/blob/5ff99e0dabefea14b04e190c1659f4817fcb7bc0/FatPkg/EnhancedFatDxe/Fat.h#L62-L82. 512b comes from the FAT spec, see section 3.1 here:  https://academy.cba.mit.edu/classes/networking_communications/SD/FAT.pdf.

```text
Count of bytes per sector. This value may take on 
only the following values: 512, 1024, 2048 or 4096.
```

Same language in exFAT: https://learn.microsoft.com/en-us/windows/win32/fileio/exfat-specification

```text
At least 9 (sector size of 512 bytes), which is the smallest sector possible for an exFAT volume
```

64kb is an implementation detail of the FAT disk cache in edk2 (see link above to Fat.h).

Reply:
Are there any checks in the FAT driver to make sure the underlying Block I/O protocol block size is one of those supported sector size values?

Reply:
That looks like a value from a structure read from the media.  Not the block size of the physical Block I/O device.  Since this PR is about caching inconsistencies, using the physical block size and not potentially some different virtual block size seems like a requirement.

Reply:
I did some digging here. We actually are using the block IO defined BlockSize to write the data cache out in this PR (see lines 695-696 in DiskCache.c). We build the cache on a virtual block size, the smallest supported by FAT, 512b. Then, when it comes time to flush the cache, we coalesce the dirty blocks into the block size supported by the device (see CacheDiskFatIo in this PR). So, I think that the only missing piece is adding an intialization check in EnhancedFatDxe/Init.c to check DeviceIo->Media->BlockSize falls in one of the buckets required by the FAT spec and fail out if not